src/policy/mallocspace/global.rs

```diff
@@ -316,9 +316,10 @@ impl<VM: VMBinding> MallocSpace<VM> {
             address,
         );
 
-        if !is_marked::<VM>(object, None) {
+        // TODO: Why do we use non-atomic load here?
+        if !unsafe { is_marked_unsafe::<VM>(object) } {
```

Comment:
I think it is a bug.  If it is not atomic, it is guaranteed to cause data race because tracing is parallel.

Reply:
I assume it should be a benign race, and it should not cause any actual bug. There are some discussions here: https://github.com/mmtk/mmtk-core/issues/313

Reply:
When @steveblackburn said "atomic operation" in https://github.com/mmtk/mmtk-core/issues/313#issuecomment-870346496, I think he meant atomic read-modify-write (RMW) operation, and atomic RMW operations do prevent duplicate edges.  In Java, all write operations to 32-bit variables and reference variables (even not `volatile`) are atomic, in the sense that it disallows word tearing, and if a read does not have happens-before relationship with a write, the read always sees some benign values (written by an actual store).  So we seldom use the phrase "atomic read" or "atomic write" in Java.

It is not true for C++ and Rust.  In C++ and Rust, if a non-atomic load has no happens-before relationship with a store, it has *undefined behaviour*.  The Rust counterpart of "ordinary" load/store in Java are `load_atomic` and `store_atomic` with the `Relaxed` order (actually the Java memory order is even weaker than Relaxed, but must prevent word tearing).

Reply:
We can change this to a relaxed atomic load, which is more reasonable.

But just for the sake of argument, I don't think this causes any actual bug. The issue #313 is clear that the `if` block may be executed multiple times for the same object. So:
* if `is_marked()` returns false, but the object is actually marked: we will just execute the block again, and it is fine.
* if `is_marked()` returns true, but the object is not actually marked: I don't think this will happen, as we are monotonically set the bit from 0 to 1 at this stage.

Reply:
I have changed this to a relaxed atomic load.